valhalla/midgard/sequence.h

```diff
@@ -321,6 +321,48 @@ template <class T> class sequence {
     std::sort(static_cast<T*>(memmap), static_cast<T*>(memmap) + memmap.size(), predicate);
     return;
   }
+  
+  // sort the file based on the predicate, and outputs to output_seq
+  //
+  // Strategy is to first sort sub-ranges of length buffer_size in place. 
+  // These should all fit in memory. Then, merge the sub-ranges into the 
+  // output sequence via priority queue.
+  void merge_sort(const std::function<bool(const T&, const T&)>& predicate,
+                  sequence<T>& output_seq,
+                  size_t buffer_size = 1024 * 1024 * 512 / sizeof(T)) {
+    flush();
+    // if no elements we are done
+    if (memmap.size() == 0) {
+      return;
+    }
+
+    // Comparator needs to be inverted for pq to provide constant time *smallest* lookup
+    // Pq keeps track of element and its index.
+    auto cmp = [&predicate](const std::pair<T, int>& a, std::pair<T, int>& b) {
+      return predicate(b.first, a.first);
+    };
+    std::priority_queue<std::pair<T, int>, std::vector<std::pair<T, int>>, decltype(cmp)> pq(cmp);
+
+    for(uint64_t i = 0; i < memmap.size(); i+=buffer_size) {
+      std::sort(static_cast<T*>(memmap) + i, 
+                static_cast<T*>(memmap) + std::min(memmap.size(), i+buffer_size), 
+                predicate);
+      pq.emplace(*at(i), i);
```

Comment:
nice! so the priority queue only stores at max however many sub-ranges it works out to based on the size of the sequence and the buffer_size :+1: